shopinvader_api_delivery_carrier/routers/delivery_carrier.py

```diff
@@ -0,0 +1,112 @@
+    data: DeliveryCarrierSearch,
+    paging_: Annotated[Paging, Depends(paging)],
+    uuid: str | None = None,
+) -> PagedCollection[DeliveryCarrierSchema]:
+    """
+    Returns the list of available carriers
+
+    If cart uuid is given or if current_cart = True in the data input,
+    the list will be limited to the
+    carriers applying to the current cart and a price will be filled
+    into the response otherwise the price is not relevant (always 0).
+
+    The field type is a technical field only use inform if the carrier
+    provides some specialized functionalities
+    """
+    cart = None
+    if uuid or data.current_cart:
+        cart = env["sale.order"]._find_open_cart(partner.id, uuid)
+
+    count, delivery_carriers = (
+        env["shopinvader_api_delivery_carrier.delivery_carrier_router.helper"]
+        .new({"partner": partner})
+        ._search(paging_, data, cart=cart)
+    )
+    return PagedCollection[DeliveryCarrierSchema](
+        count=count,
+        items=[
+            DeliveryCarrierSchema.from_delivery_carrier(delivery_carrier, cart=cart)
+            for delivery_carrier in delivery_carriers
+        ],
+    )
+
+
+class ShopinvaderApiDeliveryRouterHelper(models.AbstractModel):
+    _name = "shopinvader_api_delivery_carrier.delivery_carrier_router.helper"
+    _description = "ShopInvader API Delivery Carrier Router Helper"
+
+    partner = fields.Many2one("res.partner")
+
+    @api.model
+    def _available_carriers(self, cart):
+        cart.ensure_one()
+        sort_key = {}
+        available_carriers = cart.shopinvader_available_carrier_ids.browse()
+        for carrier in cart.shopinvader_available_carrier_ids:
+            result = carrier.rate_shipment(cart)
+            if result.get("success"):
+                available_carriers |= carrier
+                sort_key.update({carrier: result.get("price", 0.0)})
+        # Sort on the price
+        return available_carriers.sorted(lambda c: sort_key.get(c))
+
+    @property
+    def model_adapter(self) -> FilteredModelAdapter[DeliveryCarrier]:
+        return FilteredModelAdapter[DeliveryCarrier](self.env, [])
+
+    def _get(self, record_id) -> DeliveryCarrier:
+        return self.model_adapter.get(record_id)
+
+    def _search(self, paging, params, cart=None) -> tuple[int, DeliveryCarrier]:
+        if cart:
+            if params.country_id or params.zipcode:
+                ctx = self.env.context.copy()
+                ctx.update(
+                    {
+                        "delivery_force_country_id": params.country_id or 0,
+                        "delivery_force_zip_code": params.zipcode or "",
+                    }
+                )
+                cart = cart.with_context(**ctx)
+            delivery_carriers = self._available_carriers(
+                cart
+            )  # TODO: how to include paging & offset?
+            # Because even if we include order in search_with_count,
+            # we can only filter on a field...
```

Comment:
The best way for a search route is to use `self.model_adapter.search_with_count` specifying the limit and offset.
The problem is that in the case where a specific cart is given, we call `_available_carriers` that sorts the record in a specific way. This is not a simple ordering on a given field, hence it cannot be given in the `order` param of Odoo `search` method. Hence I'm not sure we can use `model_adapter.search_with_count` here. But how to deal with the paging & offset then?

Reply:
We don't keep this feature here because we don't need it. It may be added later on.

Reply:
We don't especially need paging and offset on delivery carriers, as we shouldn't have a lot of answers.